openmeter/credit/helper.go

```diff
@@ -201,46 +130,51 @@ type snapshotParams struct {
 	grants []grant.Grant
 	// Owner of the snapshot
 	owner models.NamespacedID
-	// Result of the engine.Run
-	runRes engine.RunResult
 	// Snapshot is saved if the segment is before this time & the start of the current usage period (at time of snapshot)
 	before time.Time
 }
 
 // It is assumed that there are no snapshots persisted during the length of the history (as engine.Run starts with a snapshot that should be the last valid snapshot)
-func (m *connector) snapshotEngineResult(ctx context.Context, params snapshotParams) error {
-	currentPeriodStart, err := m.ownerConnector.GetUsagePeriodStartAt(ctx, params.owner, params.runRes.Snapshot.At)
-	if err != nil {
-		return fmt.Errorf("failed to get current usage period start for owner %s at %s: %w", params.owner.ID, params.runRes.Snapshot.At, err)
-	}
-
-	segs := params.runRes.History.Segments()
+func (m *connector) snapshotEngineResult(ctx context.Context, snapParams snapshotParams, runRes engine.RunResult) error {
+	segs := runRes.History.Segments()
 
 	// i >= 1 because:
 	// The first segment starts with the last valid snapshot and we don't want to create another snapshot for that same time
 	for i := len(segs) - 1; i >= 1; i-- {
 		seg := segs[i]
 
 		// We can save a segment if its not after the current period start (this way backfilling, granting, resetting, etc... will work for the current UsagePeriod)
-		if !seg.From.After(currentPeriodStart) && !seg.From.After(params.before) {
-			snap := seg.ToSnapshot()
-			if err := m.removeInactiveGrantsFromSnapshotAt(&snap, params.grants, currentPeriodStart); err != nil {
-				return fmt.Errorf("failed to remove inactive grants from snapshot: %w", err)
+		if !seg.From.After(snapParams.before) {
+			snap, err := runRes.History.GetSnapshotAtStartOfSegment(i)
+			if err != nil {
+				return fmt.Errorf("failed to get snapshot at start of segment: %w", err)
 			}
 
-			if err := m.balanceSnapshotRepo.Save(ctx, params.owner, []balance.Snapshot{snap}); err != nil {
+			if _, err := m.saveSnapshot(ctx, snapParams, snap); err != nil {
```

Comment:
We need to be careful regarding time truncation here. I think we are fine as grants and resets are truncated, but we must never store even with the new engine such a snapshot that is not truncated to an exact second.